### PR TITLE
83 create a system to send updated player datas from client to the server

### LINKED ISSUE #83: Create a System to send updated entities to the server/clients
This System aims to select all entities that need to be modified in the other instance of the game (the server for a client/ clients for the server) and send them to the Transisthor.

### PR DESCRIPTION
**Pull request by ZQUEMA**
This Pull Request close #83 

It's add a new system to send the player's updated velocity from a client to a server

### CODE REVIEW

File: Client/Components/NetworkServer.hpp [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+{
+    ///@brief The component which will be used to link the Communicator Server'ID
+    class NetworkServer : public Component {
+      public:
+        ///@brief The id of the Communicator Server'ID
+        unsigned short id;
+
+        ///@brief Construct a new Network Server object
+        ///@param ID ID of the Communicator Server
+        NetworkServer(unsigned short ID = 0) : id(ID){};
```
Comment (LucasTesnier) [inline Client/Components/NetworkServer.hpp:22]: Kamel case for id :)

File: Client/Components/Networkable.hpp [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+    /// @brief This component is used to know which entities must be the same in the server and in client.
+    /// Each one as a unique ID.
+    class Networkable : public Component {
+      public:
+        /// @brief The Entity’s common unique identifier
+        unsigned short id;
+
+        ///@brief Construct a new Networkable object
+        ///@param ID The Entity’s common unique identifier
+        Networkable(unsigned short ID = 0) : id(ID){};
```
Comment (LucasTesnier) [inline Client/Components/Networkable.hpp:23]: kamel ;)

File: Client/Systems/SendToServer.cpp [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+    std::vector<std::shared_ptr<ecs::Entity>> servers = world.joinEntities<ecs::NetworkServer>();
+    std::vector<std::shared_ptr<ecs::Entity>> players = world.joinEntities<ecs::Controllable>();
+    std::vector<unsigned short> serverIdList;
+
+    auto addToServerList = [&serverIdList](std::shared_ptr<ecs::Entity> entityPtr) {
+        serverIdList.emplace_back(entityPtr.get()->getComponent<ecs::NetworkServer>().id);
+    };
+
+    auto update = [this, &world, &serverIdList](std::shared_ptr<ecs::Entity> entityPtr) {
+        ecs::Entity *entity = entityPtr.get();
```
Comment (LucasTesnier) [inline Client/Systems/SendToServer.cpp:21]: raw pointer ?

File: Client/Systems/SendToServer.hpp [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+** SendToServer
+*/
+
+#pragma once
+
+#include <algorithm>
+#include <concepts>
+#include <map>
+#include <typeindex>
+#include "../Components/Controllable.hpp"
```
Comment (LucasTesnier) [inline Client/Systems/SendToServer.hpp:14]: Fix in the next PR ?

File: Client/Systems/SendToServer.hpp [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+#include "../Components/NetworkServer.hpp"
+#include "../Components/Networkable.hpp"
+#include "World/World.hpp"
+
```
Comment (LucasTesnier) [inline Client/Systems/SendToServer.hpp:18]: Useless empty line

File: Client/Systems/SendToServer.hpp [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+    /// @brief A template function which is used when there is at least one component to send to the server.
+    /// @tparam C1 The first component type
+    /// @tparam ...C2 The other component types
+    /// @param world The world that the system is running in.
+    /// @param networkId The id which is common to the server and the client
+    /// @param entity Entity which must be shared
+    /// @param serverIdList The list of servers to which the datas must be sent
+    template <std::derived_from<ecs::Component> C1, std::derived_from<ecs::Component>... C2>
+    void sendToServer(ecs::World &world, const unsigned short &networkId, ecs::Entity *entity,
```
Comment (LucasTesnier) [inline Client/Systems/SendToServer.hpp:57]: Raw pointer ?

File: Client/Systems/SendToServer.hpp [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+        const std::vector<unsigned short> &serverIdList) const
+    {
+        if (componentRFCId.find(typeid(C1)) != componentRFCId.end()) {
+            if (entity->contains<C1>()) {
+                (void)serverIdList;
```
Comment (LucasTesnier) [inline Client/Systems/SendToServer.hpp:62]: Useless line

File: Client/Systems/SendToServer.hpp [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+        }
+    }
+
+    /// @brief For each player, send their velocity to the server
+    /// @param world The world that the system is running in.
+    void runSystem(ecs::World &world);
+
+    /// @brief It runs the system
+    /// @param world The world that the system is running in.
+    inline void run(ecs::World &world) override final { runSystem(world); }
```
Comment (LucasTesnier) [inline Client/Systems/SendToServer.hpp:75]: Smart, you should teach that to Nathan and Tom

File: Server/Components/NetworkClient.hpp [hunk 9]
```diff
@@ -15,11 +15,11 @@ namespace ecs
     class NetworkClient : public Component {
       public:
         ///@brief The id of the Communicator Client'ID
-        std::size_t id;
+        unsigned short id;
 
         ///@brief Construct a new Network Client object
         ///@param ID ID of the Communicator Client
-        NetworkClient(std::size_t Id = 0) : id(Id){};
+        NetworkClient(unsigned short Id = 0) : id(Id){};
```
Comment (LucasTesnier) [inline Server/Components/NetworkClient.hpp:22]: kamel :)

File: Server/Components/Networkable.hpp [hunk 10]
```diff
@@ -16,11 +16,11 @@ namespace ecs
     class Networkable : public Component {
       public:
         /// @brief The Entity’s common unique identifier
-        std::size_t id;
+        unsigned short id;
 
         ///@brief Construct a new Networkable object
         ///@param ID The Entity’s common unique identifier
-        Networkable(std::size_t Id = 0) : id(Id){};
+        Networkable(unsigned short Id = 0) : id(Id){};
```
Comment (LucasTesnier) [inline Server/Components/Networkable.hpp:23]: kamel

File: libs/ECS/World/World.hpp [hunk 11]
```diff
@@ -62,6 +62,28 @@ namespace ecs
         ///@throw std::logic_error Throw an error if the entity does not exist
         Entity &getEntity(ID id) const;
 
+        /// @brief It's a function which is used to update a component of an entity from a given distinctive component.
+        /// @tparam DistinctiveC The component type which is used to disctinct the Entity
+        /// @tparam C The component type of the component to update
+        /// @param distinctive The component which is used to disctinct the Entity
+        /// @param component The component of the component to update
+        /// @return bool True if the coponent has been modified. False otherwise.
+        template <std::derived_from<Component> DistinctiveC, std::derived_from<Component> C>
```
Comment (LucasTesnier) [inline libs/ECS/World/World.hpp:71]: Masterclass

**Review by LucasTesnier — CHANGES_REQUESTED** [review]
Some typo / doc, and question otherwise very good job man

File: Client/Systems/SendToServer.hpp (hunk 8 above)
Comment (ZQUEMA) [reply Client/Systems/SendToServer.hpp:75]: Tell us

File: Client/Systems/SendToServer.hpp (hunk 4 above)
Comment (ZQUEMA) [reply Client/Systems/SendToServer.hpp:14]: Of course

File: libs/ECS/World/World.hpp (hunk 11 above)
Comment (ZQUEMA) [reply libs/ECS/World/World.hpp:71]: Thanks !

**Review by LucasTesnier — APPROVED** [review]
NTR man !

File: Client/Components/Controllable.hpp [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+
+#include "Component/Component.hpp"
+
+namespace ecs
+{
+    /// @brief A Component which is used to know which Entity the player control
+    class Controllable : public Component {
+      public:
+        ///@brief Construct a new Controllable object
+        Controllable();
```
Comment (osvegn) [inline Client/Components/Controllable.hpp:18]: set to default

File: Client/Components/Controllable.hpp [hunk 13]
```diff
@@ -0,0 +1,23 @@
...
+
+        ///@brief Destroy the Controllable object
+        ~Controllable();
```
Comment (osvegn) [inline Client/Components/Controllable.hpp:21]: set as default

File: Client/Components/Networkable.hpp [hunk 14]
```diff
@@ -0,0 +1,29 @@
...
+*/
+
+#pragma once
+
+#include "Component/Component.hpp"
+
+namespace ecs
+{
+    /// @brief This component is used to know which entities must be the same in the server and in client.
+    /// Each one as a unique ID.
```
Comment (osvegn) [inline Client/Components/Networkable.hpp:15]: has

File: Server/Components/NetworkClient.hpp (hunk 9 above)
Comment (osvegn) [reply Server/Components/NetworkClient.hpp:22]: same as lucas (you can name it newId, or communicatorId)

File: Server/Components/Networkable.hpp (hunk 10 above)
Comment (osvegn) [reply Server/Components/Networkable.hpp:23]: same as above

**Review by osvegn — APPROVED** [review]

**Review by osvegn — CHANGES_REQUESTED** [review]
Some minor changes

**Review by LucasTesnier — DISMISSED** [review]
NTR!

**Review by LucasTesnier — APPROVED** [review]

**Review by osvegn — APPROVED** [review]

**Pull request merged**